### ISSUE TITLE
Make `error` calls GPU-friendly

### ISSUE THREAD

**Issue by charleskawczynski** [CliMA/ClimateMachine.jl#393]
There's at least one in `MoistThermo`

**Comment by vchuravy** [comment]
Errors should be okay on the GPU. As long as they don't use capture objects.

**Comment by charleskawczynski** [comment]
Ok, I'll change the title

**Comment by vchuravy** [comment]
Since Charlie just asked me to clarify what I meant:

With a captured object I mean a mutable struct where a reference to said struct is passed to a function. This can lead to that object escaping the local function and cause the object to be heap allocated. 

As an example take:

```julia
using StaticArrays
function f(i)
  A = MArray{Tuple{3}}((0,0,0))
  return A[i]
end
```

We create a mutable struct `A` that we index into. It looks like the reference to `A` does not escape the function,
but if we call it with  an out of bounds index. 

```julia
julia> f(4)
ERROR: BoundsError: attempt to access 3-element MArray{Tuple{3},Int64,1,3} with indices SOneTo(3) at index [4]
Stacktrace:
 [1] throw_boundserror(::MArray{Tuple{3},Int64,1,3}, ::Tuple{Int64}) at ./abstractarray.jl:538
 [2] checkbounds at ./abstractarray.jl:503 [inlined]
 [3] getindex at /home/vchuravy/.julia/packages/StaticArrays/3KEjZ/src/MArray.jl:91 [inlined]
 [4] f(::Int64) at ./REPL[8]:3
 [5] top-level scope at REPL[9]:1
```

We can see that there is a call to `throw_boundserror` that is not inlined and captures the reference to the `MArray`.

Inspecting the `@code_llvm` for this example, we can see the creation of a GCFrame and the heap allocation of the object.

```llvm
;  @ REPL[8]:2 within `f'
define i64 @julia_f_16107(i64) {
top:
  %gcframe = alloca %jl_value_t addrspace(10)*, i32 3
...
  %11 = call noalias nonnull %jl_value_t addrspace(10)* @jl_gc_pool_alloc(i8* %10, i32 1696, i32 32) #1
```

Julia's escape analysis could not remove the allocation, which is  something we heavily rely on for GPU compilation.

If we change the code to include an `@inbounds`
```julia
function f(i)
         A = MArray{Tuple{3}}((0,0,0))
         @inbounds return A[i]
end 
```

The GCFrame and the object allocation is gone.

```llvm
define i64 @julia_f_16145(i64) {
top:
  %1 = alloca i192, align 16
  %2 = bitcast i192* %1 to i8*
  %3 = bitcast i192* %1 to %jl_value_t*
  call void @llvm.lifetime.start.p0i8(i64 24, i8* %2)
  %4 = bitcast %jl_value_t* %3 to i8*
  call void @llvm.memset.p0i8.i64(i8* %4, i8 0, i64 24, i32 16, i1 false)
  %5 = add i64 %0, -1
  %6 = bitcast i192* %1 to i64*
  %7 = getelementptr inbounds i64, i64* %6, i64 %5
  %8 = load i64, i64* %7, align 8
  ret i64 %8
}
```

**Comment by charleskawczynski** [comment]
Very helpful, thanks! Perhaps we just need to inline moist thermo funcs!

**Comment by vchuravy** [comment]
Well `GPUifyLoops` forcefully inline's everything that is not marked at `@noinline` for that reason.Do you have specific examples of things that don't work right now? Most of the time the right thing to do is to make the errors not capture an object.

**Comment by charleskawczynski** [comment]
@vchuravy Hm, ok, maybe I'm not understanding your example then. It seemed to me like inlining was the solution to avoiding capturing an object. A really simple example is:
```
using Test, Pkg

@inline function test_err_fun(x)
    if x>0.1
        throw(ArgumentError("Bad input value"))
        return x
    else
        return x
    end
end

if haskey(Pkg.installed(), "CuArrays")
    @testset "Simple GPU test" begin
        using CuArrays
        CuArrays.allowscalar(false)
        x0 = rand(2,2)
        x = CuArray(x0)
        y = test_err_fun.(x)
    end
end
```

When running this with `julia -g2 --project` results in the stack trace 
```
ERROR: a exception was thrown during kernel execution.
Stacktrace:
ERROR: a exception was thrown during kernel execution.
Stacktrace:
ERROR: a exception was thrown during kernel execution.
Stacktrace:
ERROR: a exception was thrown during kernel execution.
Stacktrace:
 [1] test_err_fun at C:\Users\kawcz\Dropbox\Caltech\work\dev\Environments\EnvGPU\test\runtests.jl:5
 [1] test_err_fun at C:\Users\kawcz\Dropbox\Caltech\work\dev\Environments\EnvGPU\test\runtests.jl:5
 [1] test_err_fun at C:\Users\kawcz\Dropbox\Caltech\work\dev\Environments\EnvGPU\test\runtests.jl:5
 [1] test_err_fun at C:\Users\kawcz\Dropbox\Caltech\work\dev\Environments\EnvGPU\test\runtests.jl:5
 [2] _broadcast_getindex_evalf at broadcast.jl:625
 [2] _broadcast_getindex_evalf at broadcast.jl:625
 [2] _broadcast_getindex_evalf at broadcast.jl:625
 [2] _broadcast_getindex_evalf at broadcast.jl:625
 [3] _broadcast_getindex at broadcast.jl:598
 [3] _broadcast_getindex at broadcast.jl:598
 [3] _broadcast_getindex at broadcast.jl:598
 [3] _broadcast_getindex at broadcast.jl:598
 [4] getindex at broadcast.jl:558
 [4] getindex at broadcast.jl:558
 [4] getindex at broadcast.jl:558
 [4] getindex at broadcast.jl:558
 [5] #25 at C:\Users\kawcz\.julia\packages\GPUArrays\0lvhc\src\broadcast.jl:50
 [5] #25 at C:\Users\kawcz\.julia\packages\GPUArrays\0lvhc\src\broadcast.jl:50
 [5] #25 at C:\Users\kawcz\.julia\packages\GPUArrays\0lvhc\src\broadcast.jl:50
 [5] #25 at C:\Users\kawcz\.julia\packages\GPUArrays\0lvhc\src\broadcast.jl:50
Test Summary:   |
Simple GPU test | No tests
error in running finalizer: CUDAnative.KernelException(dev=CUDAdrv.CuDevice(handle=0))
```

Here, there's no mention of `"Bad input value"` nor an `ArgumentError`. However, changing `throw(ArgumentError("Bad input value"))` to `throw(ArgumentError())` changes the stack trace to

```
Simple GPU test: Error During Test at C:\Users\kawcz\Dropbox\Caltech\work\dev\Environments\EnvGPU\test\runtests.jl:18
  Got exception outside of a @test
  InvalidIRError: compiling #25(CuArrays.CuKernelState, CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(test_err_fun),Tuple{Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}) resulted in invalid LLVM IR
  Reason: unsupported dynamic function invocation (call to ArgumentError)
  Stacktrace:
   [1] test_err_fun at C:\Users\kawcz\Dropbox\Caltech\work\dev\Environments\EnvGPU\test\runtests.jl:5
   [2] _broadcast_getindex_evalf at broadcast.jl:625
   [3] _broadcast_getindex at broadcast.jl:598
   [4] getindex at broadcast.jl:558
   [5] #25 at C:\Users\kawcz\.julia\packages\GPUArrays\0lvhc\src\broadcast.jl:50
  Stacktrace:
   [1] check_ir(::CUDAnative.CompilerJob, ::LLVM.Module) at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\compiler\validation.jl:114
   [2] macro expansion at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\compiler\driver.jl:188 [inlined]
   [3] macro expansion at C:\Users\kawcz\.julia\packages\TimerOutputs\7Id5J\src\TimerOutput.jl:214 [inlined]
   [4] #codegen#152(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::typeof(CUDAnative.codegen), ::Symbol, ::CUDAnative.CompilerJob) at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\compiler\driver.jl:186
   [5] #codegen at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\compiler\driver.jl:0 [inlined]
   [6] #compile#151(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::typeof(CUDAnative.compile), ::Symbol, ::CUDAnative.CompilerJob) at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\compiler\driver.jl:47
   [7] #compile at .\none:0 [inlined]
   [8] #compile#150 at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\compiler\driver.jl:28 [inlined]
   [9] #compile at .\none:0 [inlined] (repeats 2 times)
   [10] macro expansion at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\execution.jl:403 [inlined]
   [11] #cufunction#194(::Nothing, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(CUDAnative.cufunction), ::getfield(GPUArrays, Symbol("##25#26")), ::Type{Tuple{CuArrays.CuKernelState,CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(test_err_fun),Tuple{Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}) at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\execution.jl:368
   [12] cufunction(::Function, ::Type) at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\execution.jl:368
   [13] macro expansion at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\execution.jl:176 [inlined]
   [14] macro expansion at .\gcutils.jl:87 [inlined]
   [15] macro expansion at C:\Users\kawcz\.julia\packages\CUDAnative\3Jwj2\src\execution.jl:173 [inlined]
   [16] _gpu_call(::CuArrays.CuArrayBackend, ::Function, ::CuArray{Float64,2,Nothing}, ::Tuple{CuArray{Float64,2,Nothing},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(test_err_fun),Tuple{Base.Broadcast.Extruded{CuArray{Float64,2,Nothing},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}, ::Tuple{Tuple{Int64},Tuple{Int64}}) at C:\Users\kawcz\.julia\packages\CuArrays\4ZX56\src\gpuarray_interface.jl:62
   [17] gpu_call(::Function, ::CuArray{Float64,2,Nothing}, ::Tuple{CuArray{Float64,2,Nothing},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(test_err_fun),Tuple{Base.Broadcast.Extruded{CuArray{Float64,2,Nothing},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}, ::Int64) at C:\Users\kawcz\.julia\packages\GPUArrays\0lvhc\src\abstract_gpu_interface.jl:151
   [18] gpu_call at C:\Users\kawcz\.julia\packages\GPUArrays\0lvhc\src\abstract_gpu_interface.jl:128 [inlined]
   [19] copyto! at C:\Users\kawcz\.julia\packages\GPUArrays\0lvhc\src\broadcast.jl:48 [inlined]
   [20] copyto! at .\broadcast.jl:842 [inlined]
   [21] copy(::Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(test_err_fun),Tuple{CuArray{Float64,2,Nothing}}}) at .\broadcast.jl:818
   [22] materialize(::Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Nothing,typeof(test_err_fun),Tuple{CuArray{Float64,2,Nothing}}}) at .\broadcast.jl:798
   [23] top-level scope at C:\Users\kawcz\Dropbox\Caltech\work\dev\Environments\EnvGPU\test\runtests.jl:23
   [24] top-level scope at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.2\Test\src\Test.jl:1113
   [25] top-level scope at C:\Users\kawcz\Dropbox\Caltech\work\dev\Environments\EnvGPU\test\runtests.jl:19
   [26] include at .\boot.jl:328 [inlined]
   [27] include_relative(::Module, ::String) at .\loading.jl:1094
   [28] include(::Module, ::String) at .\Base.jl:31
   [29] include(::String) at .\client.jl:431
   [30] top-level scope at none:5
   [31] eval(::Module, ::Any) at .\boot.jl:330
   [32] exec_options(::Base.JLOptions) at .\client.jl:271
   [33] _start() at .\client.jl:464
```

Now, we at least get `ArgumentError`. I don't understand how the string `"Bad input value"` in `ArgumentError` is related to captured objects but, it would be nice to be able to embed some sort of message in the exception to provide more fruitful information to users.

**Comment by simonbyrne** [comment]
> Now, we at least get `ArgumentError`. Is the string `"Bad input value"` in `ArgumentError` causing an object to be captured?

Yes, exactly. The solution is to construct the message in the `show` method for the error.

**Comment by charleskawczynski** [comment]
> The solution is to construct the message in the show method for the error.

Can you give an example?

**Comment by simonbyrne** [comment]
Hmm, actually I may be wrong. It appears that any runtime error on the GPU is catastrophic (basically you need to quit julia and restart), and you aren't going to get sensible stacktraces.

Note that the `InvalidIRError` is due to the fact that `ArgumentError()` is itself a `MethodError`, which will cause the error to be thrown at compile time instead of runtime.

**Comment by simonbyrne** [comment]
Another thing you can't do is use non-constant strings, i.e.
```
function g(x)
  if x > 0
    throw(ArgumentError("oh no"))
  end
  return x
end
```
will "work" as long as you don't hit the error branch.

On the other hand
```julia
function g(x)
  if x > 0
    throw(ArgumentError("oh no x = $x"))
  end
  return x
end
```
will throw a compile-time `ERROR: InvalidIRError` as it is trying to print the string on the GPU.

@vchuravy Since all errors are catastrophic on GPUs, could GPUifyLoops strip out any error messages on GPUs (since the user won't see them anyway).

**Comment by charleskawczynski** [comment]
Closed by #1275
